lib/web_ui/lib/src/engine/text_editing/text_editing.dart

```diff
@@ -1022,13 +1175,43 @@ abstract class DefaultTextEditingStrategy implements TextEditingStrategy {
     assert(isEnabled);
 
     final EditingState newEditingState = EditingState.fromDomElement(activeDomElement);
+    TextEditingDeltaState? newTextEditingDeltaState;
+    if (inputConfiguration.enableDeltaModel) {
+      newTextEditingDeltaState = TextEditingDeltaState.inferDeltaState(newEditingState, lastEditingState, lastTextEditingDeltaState);
+    }
 
     if (newEditingState != lastEditingState) {
       lastEditingState = newEditingState;
-      onChange!(lastEditingState);
+      lastTextEditingDeltaState = newTextEditingDeltaState;
+      onChange!(lastEditingState, lastTextEditingDeltaState);
+      // Flush delta after it has been sent to framework.
+      lastTextEditingDeltaState = TextEditingDeltaState(oldText: lastEditingState!.text!);
+    }
+  }
+
+  void handleBeforeInput(html.Event event) {
+    final String eventData = js_util.getProperty(event, 'data').toString();
```

Comment:
Maybe what you need is `js_util.getProperty(event, 'data') as String?`? That way you can avoid comparing with the string `'null'` below.